platform/android/src/style/formatted.cpp

```diff
@@ -0,0 +1,42 @@
+#include "formatted.hpp"
+#include "formatted_section.hpp"
+
+namespace mbgl {
+namespace android {
+
+void Formatted::registerNative(jni::JNIEnv& env) {
+    jni::Class<Formatted>::Singleton(env);
+}
+
+jni::Local<jni::Object<Formatted>> Formatted::New(jni::JNIEnv& env, const style::expression::Formatted& value) {
```

Comment:
This is just an idea... but this is a lot of JNI code to kind of directly wrap the constituents of `expression::Formatted` in Java equivalents, and JNI bridging code is hard to audit and reason about (including the perf implications of going back and forth across the bridge multiple times). Would it maybe make sense to try an alternative approach where the bridging code just uses `stringify(Writer& writer, const expression::Formatted& v)` to turn the `expression::Formatted` into a JSON string, uses the standard string bridging code, and then parses the JSON on the java side? It's not at all clear to me this would be a win in terms of clarity or performance, but thought it might be worth considering...

Reply:
JSON parsing has been a bottleneck on multiple occasions and I feel like JNI overhead might be the lesser of two evils in this case.

Reply:
OK, fair enough. Since we're now doing this even for the string-typed `getTextField` (since the format->string flattening happens in java-land now), this could affect performance for existing code. Do you have an idea what an appropriate benchmark for this would be? I don't have a good feel for how often users will typically be hitting `getTextField`, it totally makes sense that it might not be nearly heavy enough to matter.

Reply:
I did a quick test - 1000 subsequent calls to `getTextField`:
- `master` ~6ms
- this PR ~21ms

The overhead is significant overall, but I don't think it's significant in this particular scenario. Constant `text-field` is definitely not a property that's going to be queried a lot as it depends on source updates that need to be performed by the user, as do all constant properties. Also, I still believe that JSON parsing on the java side would introduce even bigger performance hit.

Reply:
Sounds good. A ~4x slowdown doesn't sound disastrous in this context at all.